{% extends 'base_wt.html' %}


{% block title %} | Neighborhood Analyzer{% endblock %}

{% block content %}

<main role="main">
    <form action = "{% url 'count_from_domain_family' %}" >
        <div class="container">
          <h1 class="text-center pt-3">Neighbourhood Analyzer - PFAM Domain (Family)</h1>
          <br/>
          <br/>
          <p>	
		In prokaryotic genomes, functionally coupled genes can be organized in conserved gene clusters, e.g. operons 
		enabling their coordinated regulation. Thus, it is possible to predict function of uncharacterised genes 
		by analyzing functional annotations within their neighborhoods. Here, we present an algorithm that gives an insight 
		into genomic neighborhoods of a query protein family by calculating statistical significance (Wilcoxon test)	
		for overrepresentation of functional domains in the neighbourhood. 
	   </p>
	   <p>
		&raquo;As query, the user can choose a protein family from the Pfam database (September 2018, 17929 entries),
		<a href="http://www.pfam.xfam.org">PFAM MAIN SITE</a>. For input, query domain has to be spelled "pfNNNNN" or "PFNNNNN",
		e.g. PF02696.
	   <br>	
		&raquo; The size of neighbourhood is described as number of basepairs before and after the gene whose protein 
		product possesses the query pfam domain. Typically, the size of neighborhood set to 5000 bp corresponds to 10 genes 
		in whole neighborhood.
		
	   <br>
		&raquo; Due to possibility of high false discovery rate that can be a serious problem when multiple tests are performed 
		(e.g. many Pfam domains are evaluated for overrepresentation), the  user can decide which method of multiple test correction to use. 
	   <br>
		<a href="https://en.wikipedia.org/wiki/Bonferroni_correction">Bonferroni correction</a> or 
		<a href="http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.829.8800&rep=rep1&type=pdf">Benjamini-Hochberg procedure</a>
	   <br>
		&raquo; Families that are available for neighborhood analysis were chosen based on high number of representative genomes available.
		The genomic data that is used in this server came from <a href="https://img.jgi.doe.gov/">
		JGI IMG Integrated Microbial Genomes & Microbiomes</a> and it contains prokaryotic complete genomes and 
		uncomplete data from environmental sequencing.
	   <br>
		&raquo; Currently, during neighborhood analysis both strands are taken to collect information about protein 
		families. However, in the future it will be possible also to choose only the strand on which the query pfam 
		domain is present in the analysed genome. 
	   <br>
		&raquo; Cut-off value is used to narrow the output based on p-value, and name of analysis is just for user's convenience 

	   </p>
          <br/>
          <br/>
    <div class="alert alert-danger" role="alert">
        <h4 class="alert-heading">Wait a minute !</h4>
        <p>
            This tool is our newest invention so feel feel free to use it and test in many different ways. Before
            using it please read carefully instructions above.

        </p>
        <hr>
        <p class="mb-0">If you have any suggestions or ideas that will help us improve our tool, feel free to mail us ;)</p>
    </div>





    <form>
      <div class="form-row">
        <div class="form-group col-md-6">
          <label for="PfamDomain">Which protein domain you want to analyse?</label>
          <input type="text" class="form-control" name = "pfam_domain" placeholder="PF02696"  required="">
        </div>
      </div>
      <div class="form-row">
        <div class="form-group col-md-6">
          <label for="Range">Set size of neighbourhood, you want to investigate (bp for up and downstream)</label>
          <input class="form-control" name ="range_search" value="5000" required="" type="number"  min="0" max="20000" step="1">
        </div>
      </div>
      <div class="form-row">
        <div class="form-group col-md-6">
          <label for="MultiTestCorrection">Choose which multiple test correction method you want to use</label>
          <select name="test_correction" class="form-control">
            <option selected value='none'>None</option>
            <option value='bonferroni'>Bonferroni Correction</option>
            <option value='fdr_bh'>Benjamini-Hochberg Procedure</option>
          </select>
        </div>
      </div>
      <div class="form-row">
        <div class="form-group col-md-6">
          <label for="Taxa">Choose which family you want to investigate</label>
          <select name="database_taxa" class="form-control">

            <option value="Enterobacteriaceae">Enterobacteriaceae (44 genera --> 9571 genomes)</option>
            <option value="Staphylococcaceae">Staphylococcaceae (8 genera --> 5904 genomes)</option>
            <option value="Mycobacteriaceae">Mycobacteriaceae (7 genera --> 4332 genomes)</option>
            <option value="Streptococcaceae">Streptococcaceae (3 genera --> 3112 genomes)</option>
            <option value="Pseudomonadaceae">Pseudomonadaceae (8 genera --> 2380 genomes)</option>
            <option value="Bacillaceae">Bacillaceae (59 genera --> 2112 genomes)</option>
            <option value="Vibrionaceae">Vibrionaceae (9 genera --> 1801 genomes) </option>
            <option value="Moraxellaceae">Moraxellaceae (8 genera --> 1701 genomes)</option>
            <option value="Streptomycetaceae">Streptomycetaceae (5 genera --> 1605 genomes)</option>
            <option value="Burkholderiaceae">Burkholderiaceae (17 genera --> 1457 genomes)</option>
            <option value="Flavobacteriaceae">Flavobacteriaceae (94 genera --> 1051 genomes)</option>
            <option value="Lactobacillaceae">Lactobacillaceae (4 genera --> 1038 genomes)</option>

          </select>
        </div>
      </div>
      <div class="form-row">
        <div class="form-group col-md-6">
          <label for="Strand">Choose which DNA strand you are insterested in </label>
          <select name="strand_select" class="form-control">
            <option selected value="both">Both strands</option>
            <!-- FOR FUTURE SCRIPT!!!
            <option value="same">Same as searched pfam</option>
            <option value="oposite">Oposite as searched pfam</option>
            -->
          </select>
        </div>
      </div>
      <div class="form-row">
        <div class="form-group col-md-6">
          <label for="CutOff">Set cut-off value for output from statistic test </label>
          <select name="cut_off" class="form-control">
            <option selected value="none">None</option>
            <option value="0">Only zeros</option>
            <option value="0.00001">0.00001</option>
            <option value="0.00005">0.00005</option>
            <option value="0.0001">0.0001</option>
            <option value="0.0005">0.0005</option>
            <option value="0.001">0.001</option>
            <option value="0.005">0.005</option>
            <option value="0.01">0.01</option>
            <option value="0.05">0.05</option>
            <option value="0.1">0.1</option>
            <option value="0.2">0.2</option>
            <option value="0.5">0.5</option>
          </select>
        </div>
      </div>
      <div class="form-row">
        <div class="form-group col-md-6">
          <label for="PfamDomain">Give a name to your analysis</label>
          <input type="text" class="form-control" name = "out_name" placeholder="42 is the answer"  required="">
        </div>
      </div>
      <br/>
      <br/>
      {% if user.is_authenticated %}
        <input type="submit" name="analyze" value="Analyze ! "  class="btn btn-primary my-2" >
      {% else %}
        <a href="{% url 'login' %}" class="btn-group">
          <button type="button" class="btn btn-primary my-2 ">Login first !</button>
        </a>
      {% endif %}  

      
    </form>
    <br/>
    <br/>
    <br/>
    <br/>
    <br/>
    <br/>
    <br/>
    <br/>
    <br/>
    <br/>
    <br/>
    <br/>
  </main>


{% endblock %}